### PR TITLE
Improve static cached_op optimization

### PR DESCRIPTION
**Pull request by xinyu-intel**
fix along with #14931

@pengzhao-intel @ZhennanQin 

## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/imperative/imperative_utils.h [hunk 1]
```diff
@@ -595,7 +595,10 @@ inline bool CheckAndInferShape(nnvm::Graph* p_g, mxnet::ShapeVector&& shapes,
     *contain_unknown = false;
   }
   nnvm::Graph& g = *p_g;
-  if (g.attrs.count("shape")) {
+  if (use_inputs) {
```
Comment (pengzhao-intel) [inline src/imperative/imperative_utils.h]: Please add the comments in the code

**Comment by piyushghai** [comment]
Thanks for your contributions @xinyu-intel .
Can you also look into the CI failures ? 

@mxnet-label-bot  Add [pr-awaiting-review, Operator, Backend]

**Comment by pengzhao-intel** [comment]
Please also try https://github.com/dmlc/gluon-nlp/issues/690

**Comment by ZhennanQin** [comment]
@pengzhao-intel I've tested https://github.com/dmlc/gluon-nlp/issues/690, and everything works fine.

File: src/imperative/imperative_utils.h (hunk 1 above)
Comment (ZhennanQin) [reply src/imperative/imperative_utils.h]: This code is removed from previous commit by mistake. This is just to recover them back.

**Comment by pengzhao-intel** [comment]
> @pengzhao-intel I've tested [dmlc/gluon-nlp#690](https://github.com/dmlc/gluon-nlp/issues/690), and everything works fine.

Thanks :)

File: src/imperative/cached_op.cc [hunk 2]
```diff
@@ -290,7 +290,7 @@ bool CachedOp::CheckDynamicShapeExists(const Context& default_ctx,
   CheckAndInferShape(&g, std::move(shape_inputs), true,
                      {0, 0}, {0, 0},
                      &contain_dynamic_shape);
-  if (contain_dynamic_shape && erase_result) {
+  if (!config_.static_shape && erase_result) {
```
Comment (marcoabreu) [inline src/imperative/cached_op.cc]: Sounds like something wasn't properly tested here? Maybe add a test to cover this path?

Comment (ZhennanQin) [reply src/imperative/cached_op.cc]: It's already covered by current UT, there're 2 cases failed before fixing this.

**Comment by roywei** [comment]
@PatricZhao @ZhennanQin  Is this good to go? 
We recently observed static alloc regression on some sue cases. One of the case is reported here https://github.com/apache/incubator-mxnet/issues/15067. 
another is here: https://lists.apache.org/thread.html/7263d1be69d7e6f07b81c28fc1d209f30bdd3d36ea860ebbffddf96d@%3Cdev.mxnet.apache.org%3E

Although we are not sure if this fix is related, can we add this to 1.5.0 anyway since #14931 is already in 1.5.0.rc0 ? 
The vote is cancelled, and pending fix for horovod fix. So we have some time to merge this.
Thanks!

**Comment by pengzhao-intel** [comment]
> @PatricZhao @ZhennanQin Is this good to go?
> We recently observed static alloc regression on some sue cases. One of the case is reported here #15067.
> another is here: https://lists.apache.org/thread.html/7263d1be69d7e6f07b81c28fc1d209f30bdd3d36ea860ebbffddf96d@%3Cdev.mxnet.apache.org%3E
> 
> Although we are not sure if this fix is related, can we add this to 1.5.0 anyway since #14931 is already in 1.5.0.rc0 ?
> The vote is cancelled, and pending fix for horovod build. So we have some time to merge this.
> Thanks!

Agree! I think the PR is ready to merge.

**Comment by pengzhao-intel** [comment]
@roywei merged, please take it into 1.5 and try if the regression issue is fixed.
Feel free to ping us if anything needs our help.

**Pull request merged**